aidb/vector_database/marqo_vector_database.py

```diff
@@ -0,0 +1,223 @@
+                        'space_type': similarity
+                    }
+                }
+            }
+      )
+    self.index_list.append(index_name)
+
+
+  def load_index(self):
+    '''
+    Reload index from Marqo
+    '''
+    self.index_list = [i.index_name for i in self.marqo_client.get_indexes()['results']]
+
+
+  def delete_index(self, index_name: str):
+    '''
+    delete an index
+    '''
+    if index_name in self.index_list:
+      self.marqo_client.delete_index(index_name)
+      # self.index_list.remove(index_name)
+      self.index_list = [i.index_name for i in self.marqo_client.get_indexes()['results']]
+      logger.info("Index '%s' deleted.", index_name)
+  
+
+  def _check_index_validity(self, index_name: str):
+    if index_name not in self.index_list:
+      raise Exception(f'Couldn\'t find index {index_name}, please create it first')
+    return index_name
+  
+
+  def insert_data(self, index_name: str, data: pd.DataFrame):
+    '''
+    insert data into an index
+    '''
+    index_name = self._check_index_validity(index_name)
+
+    data = data.to_dict(orient="records")
+    # Data will have list of dictionaries 
+    # with 'id' and 'values' as keys.
+    
+    mod_data = []
+    for d in data:
+      # print(d)
+      _id = str(d['id'])
+      _vec = d['values']
+      mod_data.append(
+        {
+          '_id': _id,
+          'aidb_data':{
+            'vector': _vec
+          }
+        }
+      )
+    response = self.marqo_client.index(
+        index_name=index_name
+        ).add_documents(
+          documents=mod_data,
+          mappings={
+            'aidb_data':{
+              'type': 'custom_vector'
+            }
+          },
+          tensor_fields=['aidb_data'],
+          auto_refresh=True,
+          client_batch_size=64
+        )
+    
+
+  def get_embeddings_by_id(self, index_name: str, ids: np.ndarray, reload = False) -> np.ndarray:
+    '''
+    Get data by id and return results
+    '''
+    if reload:
+      self.load_index()
+    index_name = self._check_index_validity(index_name)
+    result = []
+    resp = self.marqo_client.index(
+        index_name).get_documents(
+          document_ids=[str(i) for i in ids], expose_facets=True
+        )
+    result = [ii['_tensor_facets'][0]['_embedding'] for ii in resp['results']]
+    return np.array(result)
+  
+
+  def query_by_embedding(
+      self,
+      index_name: str,
+      query_emb_list: np.ndarray,
+      top_k: int = 5,
+  ) -> (np.ndarray, np.ndarray):
+    '''
+    Query nearest k embeddings, return embeddings and ids
+    '''
+    index_name = self._check_index_validity(index_name)
+    
+    all_topk_reps, all_topk_dists = [], []
+    for query_emb in query_emb_list:
+      response = self.marqo_client.index(index_name).search(
+        # No query needs to be provided when no_model and custom vectors,
+        # only context - https://docs.marqo.ai/1.4.0/API-Reference/Indexes/create_index/#no-model
```

Comment:
Put comments before the corresponding code line.

Reply:
Done